File: python/leetcode/164.py
class Solution:
    def maximumGap(self, num):
        if len(num) < 2 or min(num) == max(num): return 0
        a, b = min(num), max(num)
        size = math.ceil((b-a)/(len(num)-1))
        bucket = [[None, None] for _ in range((b-a)//size+1)]
        for n in num:
            b = bucket[(n-a)//size]
            b[0] = n if not b[0] else min(b[0], n)
            b[1] = n if not b[1] else max(b[1], n)
        bucket = [b for b in bucket if b[0]]
        return max(bucket[i][0] - bucket[i-1][1] for i in range(1, len(bucket)))

File: cpp/leetcode/164.cpp
class Solution {
public:
    int maximumGap(vector<int>& nums) {
        if(nums.empty()) return 0;
        int pmax = INT_MIN, pmin = INT_MAX, n = nums.size();
        for(auto num : nums){
            pmax = max(num, pmax);
            pmin = min(num, pmin);
        }
        int width = (pmax - pmin) / n + 1;
        int size = (pmax - pmin) / width + 1;
        vector<int> dmax(size, INT_MIN), dmin(size, INT_MAX);
        unordered_set<int> s;
        for(auto num : nums){
            int idx = (num - pmin) / width;
            dmax[idx] = max(dmax[idx], num);
            dmin[idx] = min(dmin[idx], num);
            s.insert(idx);
        }
        int res = 0, pre = 0;
        for(int i = 1; i < size; i++){
            if(s.find(i) == s.end()) continue;
            res = max(res, dmin[i] - dmax[pre]);
            pre = i;
        }
        return res;
    }
};